### ISSUE TITLE
get_dynamo_library_bounds sometimes does not correctly find bounds

### ISSUE THREAD

**Issue by Carrotman42** [DynamoRIO/dynamorio#3387]
Internally we have a heisenbug which is causing a hang on static detach via `dr_app_stop_and_cleanup_with_stats`. (Given a fixed compilation/configuration for a specific application the bug is always or never there, but we do not know exactly what compiler feature/runtime configuration is causing the bug to show up: the bug keeps showing up and disappearing depending on where we are sync'd in the source history.) This bug tracks its resolution in relation to DR core code.

I was able to show that when there is a detach hang, `get_dynamo_library_bounds` is not "correctly" discovering the full bounds of DR code; that is, in one instance the bounds were found to be [`0x13200000`, `0x1d600000`) whereas `dr_app_stop_and_cleanup_with_stats` had an address of `0x131eb110`.

It is unclear at this point if the bug is truly related to the code layout of the binary (e.g. the layout algorithms just so happen to place dr_app_stop_and_cleanup_with_stats far away from all other DR code) or if it's related to a bug in `get_dynamo_library_bounds` (e.g. not correctly handling some edge case in its scanning and heuristics).

xref PR#3359, which was also related to debugging this issue.

**Comment by derekbruening** [comment]
What we want is the contents of /proc/self/maps for that `[0x13200000, 0x1d600000)` case.  My theory is that this is related to hugepage anonymous mremaps: xref #2566.

**Comment by derekbruening** [comment]
> It is unclear at this point if the bug is truly related to the code layout of the binary (e.g. the layout algorithms just so happen to place dr_app_stop_and_cleanup_with_stats far away from all other DR code) or if it's related to a bug in get_dynamo_library_bounds 

Not sure I see what's unclear?  If there are symbols in the app module that are outside the bounds of that module as found by DR, it seems to clearly be a bug in `get_dynamo_library_bounds`.

**Comment by Carrotman42** [comment]
As an update: the issue here is that `memquery_library_bounds_by_iterator` does not correctly handle our internal mapping setup. Specifically, while that function contains logic to join consecutive mappings in /proc/self/maps into the same "library" up if they have the same comment field, internally these comments are not always the same. This causes the logic to work sometimes and other times not, depending on compile-time and run-time configurations.

Derek and I discussed utilizing the existing whole-address-space walk from `find_dynamo_library_vm_areas` to implement `get_dynamo_library_bounds` instead of trying to infer the bounds from an example PC (for static DR), but unfortunately `get_dynamo_library_bounds` needs to be called much earlier than when `find_dynamo_library_vm_areas` is first called (which is around the time vmarea is initialized afaict).

Another option would be to change the logic of `memquery_library_bounds_by_iterator` to proactively look for ELF headers and the information therein to understand the layout of the mappings. There are a few comments in the source about not trusting these headers, though. Quoting Derek from another medium:

    I think the two problems with relying on the ELF header are:

    1) Later segments sometimes include the 1st page, which I believe is the
         typical source of the extra headers mid-library
    2) Sometimes ELF files are mapped just for data, so just because there's a
         header doesn't mean the segments are mapped instead of just the file

Through some research I believe we can look through the ELF headers to calculate the expected segment mappings and match those with the maps file, but I fear it will just be another heuristic: for late-attach use cases, an application probably can arbitrarily remap itself however it wants and I'm not sure we can definitively tell where the application starts and stops. (I'm sure we can't even guarantee the application will be mapped contiguously in all cases.)

**Comment by derekbruening** [comment]
> (I'm sure we can't even guarantee the application will be mapped contiguously in all cases.)

Yes, there can certainly be gaps in general in ELF files into which other things can be mapped: we have seen this numerous times.  However, we ensure this never happens for libdynamorio.so (https://github.com/DynamoRIO/dynamorio/blob/master/core/unix/loader.c#L1855) or for the app itself when DR launches it (our private loader leaves an mmap there).

Unfortunately for static DR those safeguards are not in place.
